### PR TITLE
Beatloops snap to the closest (fractional) beat in quantized mode LP1752133

### PR DESCRIPTION
**Pull request by goddisignz**
Quantized beat loops always started on the previous beat while loop in and loop out markers snapped to the closest beat. Now both snap to the closest beat. If the closest beat is ahead of the current play position a catching loop is generated.

Fixes https://bugs.launchpad.net/mixxx/+bug/1752133

### CODE REVIEW

File: src/engine/controls/loopingcontrol.cpp [hunk 1]
```diff
@@ -1078,13 +1078,28 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
...
                 // If I press 1/4 beatloop, we want loop from 50% to 75% etc
-                double beat_len = nextBeat - prevBeat;
-                double loops_per_beat = 1.0 / beats;
-                double beat_pos = currentSample - prevBeat;
-                int beat_frac =
-                        static_cast<int>(floor((beat_pos / beat_len) *
-                                                loops_per_beat));
-                newloopSamples.start = prevBeat + beat_len / loops_per_beat * beat_frac;
+                double beatLength = nextBeat - prevBeat;
+                double beatPos = currentSample - prevBeat;
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: this can be samplesSinceBeat or so

File: src/engine/controls/loopingcontrol.cpp [hunk 2]
```diff
@@ -1078,13 +1078,28 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
...
+
+                // find the two closest beat fractions and place the new loop start to the closer one
+                double fractionBeatLength = beatLength * beats;
+                double previousFractionBeat = prevBeat + (floor(beatPos / fractionBeatLength)) * fractionBeatLength;
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: The () around is redundant. You can use it to emphasize that * is done first or move the + prevBeat to the end.

File: src/engine/controls/loopingcontrol.cpp [hunk 3]
```diff
@@ -1078,13 +1078,28 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
...
+                double nextFractionBeat = prevBeat + (floor(beatPos / fractionBeatLength) + 1) * fractionBeatLength;
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: double nextFractionBeat = previousFractionBeat + fractionBeatLength;
:-)

File: src/engine/controls/loopingcontrol.cpp [hunk 4]
```diff
@@ -1078,13 +1078,28 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
...
+
+                if (abs(previousFractionBeat - currentSample) <= abs(nextFractionBeat - currentSample)) {
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: if (currentSample - previousFractionBeat <= nextFractionBeat - currentSample) {

File: src/engine/controls/loopingcontrol.cpp [hunk 5]
```diff
@@ -1078,13 +1078,28 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
...
+                } else {
+                    newloopSamples.start = nextFractionBeat;
+                }
+
+                // If running reverse, move the loop one fractional beat to the left.
+                // Thus, the loop end is (time-wise) in front of the current position.
+                //
+                // TODO: There must be a better way of checking if we are running reverse. Getting the "old_reverse" from the EngineBuffer somehow seems wrong.
+                // Furthermore, it does not work when we are not playing and the reverse button is held.
+                if (getEngineBuffer()->isReverse()) {
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: For now I have now better idea. This is the real reverse state of the last engine call that includes scratching and seeks. You actually want the future. This: ControlPushButton(ConfigKey(group, "reverse")); would be the reverse button. 
This excludes scratches and seeks, probably desired?

File: src/track/trackmetadatataglib.cpp [hunk 6]
```diff
@@ -1867,7 +1867,7 @@ bool exportTrackMetadataIntoID3v2Tag(TagLib::ID3v2::Tag* pTag,
     DEBUG_ASSERT(pHeader);
     if (!checkID3v2HeaderVersionSupported(*pHeader)) {
         kLogger.warning() << "Legacy ID3v2 version - exporting only basic tags";
-        exportTrackMetadataIntoTag(pTag, trackMetadata);
+        exportTrackMetadataIntoTag(pTag, trackMetadata, WRITE_TAG_OMIT_NONE);
```
Comment (daschuer) [inline src/track/trackmetadatataglib.cpp]: unrelated change?

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for continuing this. I have left some comments.

File: src/track/trackmetadatataglib.cpp (hunk 6 above)
Comment (goddisignz) [reply src/track/trackmetadatataglib.cpp]: Yep, totally unlerated but building failed without this. Occurred since one master merge from @uklotzde.

Comment (daschuer) [reply src/track/trackmetadatataglib.cpp]: Ah, I see this change is already part of current master. 
Can you rebase your commit on top of the current master: 
"git rebase upstream/master" 
This way we avoid to track the same change in two git branches.

Comment (goddisignz) [reply src/track/trackmetadatataglib.cpp]: Worked perfectly.

**Comment by daschuer** [comment]
Thank you for this PR and the changes. LGTM
